pkg/cvo/metrics.go

```diff
@@ -449,3 +552,85 @@ func mostRecentTimestamp(cv *configv1.ClusterVersion) int64 {
 	}
 	return latest.Unix()
 }
+
+// Determine if the certificates have changed and need to be updated.
+// If no errors occur, returns true if both files have changed and
+// neither is an empty file. Otherwise returns false and any error.
+func certsChanged(origCertChecksum []byte, origKeyChecksum []byte, certFile, keyFile string) (bool, error) {
+	// Check if both files exist.
+	certNotEmpty, err := fileExistsAndNotEmpty(certFile)
+	if err != nil {
+		return false, fmt.Errorf("Error checking if changed TLS cert file empty/exists: %w", err)
+	}
+	keyNotEmpty, err := fileExistsAndNotEmpty(keyFile)
+	if err != nil {
+		return false, fmt.Errorf("Error checking if changed TLS key file empty/exists: %w", err)
+	}
+
+	if !certNotEmpty || !keyNotEmpty {
+		// One of the files is missing despite some file event.
+		return false, fmt.Errorf("Certificate or key is missing or empty, certificates will not be rotated.")
+	}
+	currentCertChecksum := checksumFile(certFile)
+	currentKeyChecksum := checksumFile(keyFile)
+
+	// Check if the non-empty certificate/key files have actually changed.
+	if !bytes.Equal(origCertChecksum, currentCertChecksum) && !bytes.Equal(origKeyChecksum, currentKeyChecksum) {
+		klog.V(2).Info("Certificate and key changed. Will recreate metrics server with updated TLS configuration.")
+		return true, nil
+	}
+
+	return false, nil
+}
+
+func makeTLSConfig(servingCertFile, servingKeyFile string) (*tls.Config, error) {
+	// Load the initial certificate contents.
+	certBytes, err := ioutil.ReadFile(servingCertFile)
+	if err != nil {
+		return nil, err
+	}
+	keyBytes, err := ioutil.ReadFile(servingKeyFile)
+	if err != nil {
+		return nil, err
+	}
+	certificate, err := tls.X509KeyPair(certBytes, keyBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return crypto.SecureTLSConfig(&tls.Config{
+		GetCertificate: func(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
+			return &certificate, nil
+		},
+	}), nil
+}
+
+// Compute the sha256 checksum for file 'fName'.
+func checksumFile(fName string) []byte {
+	file, err := os.Open(fName)
+	if err != nil {
+		klog.Errorf("Failed to open file %v for checksum: %v", fName, err)
```

Comment:
Per [my pitch for crashlooping when we have trouble with things that should always succeed][1], can we make `Open` and `Copy` failures fatal and return an `error` from this function?

[1]: https://github.com/openshift/cluster-version-operator/pull/726#discussion_r786276935

Reply:
As of 0e6136c727b055, `checksumFile` returns an `error` too.